### PR TITLE
Fixes #6488: Avoid waiting on lock when calling yum

### PR DESCRIPTION
**Pull request by ncharles**
https://www.rudder-project.org/redmine/issues/6488

### CODE REVIEW

File: tree/20_cfe_basics/cfengine/packages.cf [hunk 1]
```diff
@@ -99,6 +99,19 @@ bundle common rpm_knowledge
       "rpm3_arch_regex" string => "\S+\s+(\S+).*";
 }
 
+bundle common redhat_no_locking_knowledge {
+  vars:
+    !redhat_setopt_not_available::
+      "no_locking_command" string => "--setopt=exit_on_lock=True";
+    redhat_setopt_not_available::
+      "no_locking_command" string => "";
```
Comment (jooooooon) [inline tree/20_cfe_basics/cfengine/packages.cf]: I know this is nitpicking, but it's really not a *command*, it's an *option*.

File: tree/20_cfe_basics/cfengine/packages.cf [hunk 2]
```diff
@@ -99,6 +99,19 @@ bundle common rpm_knowledge
       "rpm3_arch_regex" string => "\S+\s+(\S+).*";
 }
 
+bundle common redhat_no_locking_knowledge {
+  vars:
+    !redhat_setopt_not_available::
```
Comment (jooooooon) [inline tree/20_cfe_basics/cfengine/packages.cf]: This class is sensitive to normal ordering. In my tests, the result of the real test below didn't matter, this case was always set. I strongly recommend adding a "check_done" class below and prefixing it to this condition.

File: tree/20_cfe_basics/cfengine/packages.cf [hunk 3]
```diff
@@ -99,6 +99,19 @@ bundle common rpm_knowledge
...
+      "no_locking_command" string => "--setopt=exit_on_lock=True";
+    redhat_setopt_not_available::
+      "no_locking_command" string => "";
+
+
+  classes:
+    redhat::
+      "redhat_setopt_not_available" expression => returnszero("LANG=C ${paths.path[yum]} --setopt=exit_on_lock=True 2>&1 | ${paths.path[grep]} -E '(Command line error: no such option: --setopt|Main config did not have a exit_on_lock attr. before setopt)'", "useshell");
```
Comment (jooooooon) [inline tree/20_cfe_basics/cfengine/packages.cf]: This returns a false positive when yum is simply not available.

I propose reversing the test:

` "redhat_setopt_available" not => returnszero("LANG=C ${paths.path[yum]} --setopt=exit_on_lock=True 2>&1 | ${paths.path[grep]} -E '(Command line error: no such option: --setopt|Main config did not have a exit_on_lock attr. before setopt)'", "useshell"); `

File: tree/20_cfe_basics/cfengine/packages.cf (hunk 2 above)
Comment (ncharles) [inline tree/20_cfe_basics/cfengine/packages.cf]: classes in bundle common are global - so i'm not super happy to add extra classes there that will leak in global context and might overlap

File: tree/20_cfe_basics/cfengine/packages.cf (hunk 3 above)
Comment (ncharles) [inline tree/20_cfe_basics/cfengine/packages.cf]: this option is only used in conjonction with yum binary, and checked on redhat only.
I agree that reversing the condition will make it clearer

**Comment by ncharles** [comment]
Commit modified

**Comment by jooooooon** [comment]
Looks good to me, merging. We should also propose this fix upstream.

**Pull request merged**